jlm/util/AnnotationMap.hpp

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2024 Nico Reißmann <nico.reissmann@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_UTIL_ANNOTATION_MAP_HPP
+#define JLM_UTIL_ANNOTATION_MAP_HPP
+
+#include <jlm/util/common.hpp>
+#include <jlm/util/iterator_range.hpp>
+
+#include <cstdint>
+#include <string_view>
+#include <unordered_map>
+#include <variant>
+
+namespace jlm::util
+{
+
+/**
+ * Represents a simple key-value pair with a label and a value of type std::string_view, int64_t,
+ * uint64_t, or double.
+ */
+class Annotation final
+{
+  using AnnotationValue = std::variant<std::string_view, int64_t, uint64_t, double>;
+
+public:
+  Annotation(std::string_view label, std::string_view value)
+      : Label_(std::move(label)),
+        Value_(std::move(value))
+  {}
+
+  Annotation(std::string_view label, int64_t value)
+      : Label_(std::move(label)),
+        Value_(std::move(value))
+  {}
+
+  Annotation(std::string_view label, uint64_t value)
+      : Label_(std::move(label)),
+        Value_(std::move(value))
+  {}
+
+  Annotation(std::string_view label, double value)
+      : Label_(std::move(label)),
+        Value_(std::move(value))
+  {}
+
+  /**
+   * Gets the label of the annotation.
+   */
+  [[nodiscard]] const std::string_view &
+  Label() const noexcept
+  {
+    return Label_;
+  }
+
+  /**
+   * Gets the value of the annotation. Requires the annotation value to be of type \p TValue.
+   */
+  template<typename TValue>
+  [[nodiscard]] const TValue &
+  Value() const
+  {
+    return std::get<TValue>(Value_);
+  }
+
+  /**
+   * Checks if the type of the annotation value is equivalent to \p TValue.
+   *
+   * @return True if the value type if equivalent to \p TValue, otherwise false.
+   */
+  template<typename TValue>
+  [[nodiscard]] bool
+  HasValueType() const noexcept
+  {
+    return std::get_if<TValue>(&Value_) != nullptr;
```

Comment:
`std::holds_alternative` is maybe nicer?

Reply:
It is nicer. Changed!